### DISCUSSION TITLE
how to get latest emitted value from a observable

### DISCUSSION THREAD

**Discussion by xianshenglu** [ReactiveX/rxjs#6887] [Ideas / Feature request]
```ts
const { combineLatest, interval, of } = rxjs;
const { first, last, sample, take, withLatestFrom } = rxjs.operators;

const numbers = interval(1000);

const takeFourNumbers = numbers.pipe(take(4));
takeFourNumbers.subscribe(x => console.log('Next: ', x));
setTimeout(()=>{
 console.log('how can we get the latest value which is 1?');
 takeFourNumbers.pipe(first()).subscribe(v=>console.log(v,'actual get'))
},2500)
// Logs:
// Next: 0
// Next: 1
// how can we get the latest value which is 1?
// Next: 2
// 0 actual get
// Next: 3
```
As the demo shows, as the title said, how to get the latest emitted value from an observable?

In my case, I just want to get the latest value and do a one-time job. How can I do it?

Maybe we can have an operator like `takeLatest()` or `latest()`?

**Comment by christophka** [comment]
I think that there are ways already available for your use case, although the best way depends on the details. On the other hand I think it might be useful to know that in your example the reason why the output is `0 actual get` has to do with how `Observable`s work. Loosely speaking, subscribing to the same observable twice is like calling the same function twice. An `Observable` produces values (e.g. the `Observable` created with calling `interval(1000)` will produce increasing values every second), much like a function like this
```ts
function answer() { return 42; }
```
produces a single value. Calling this function twice will produce the same value. The same is true for `Observable`s, so that in your case when you subscribe the second time with
```ts
takeFourNumbers.pipe(first()).subscribe(v=>console.log(v,'actual get'))
```
what you're doing is to start producing the values as "defined" by how `interval(1000).pipe(take(4))` will produce them. When I started to work with RxJS I had a hard time to wrap my head around that, until I wrote my first "pure" `Observable` (with `new Observable(...)`) which helped me understand the concept a lot better. Imho this is pretty good summed up on the [RxJS doc on Observables](https://rxjs.dev/guide/observable#observables-as-generalizations-of-functions).

Going back to your original question, there are several options now that depend on your use case.

## Imperatively
I'd use this option when you're in an imperative use case where you need to have the latest value. The simplest way would be to just store the last emitted value
either from your subscription
```ts
takeFourNumbers.subscribe(x => {
   ...
   latestValue = x;
});
```
or using the `tap` operator
```ts
takeFourNumbers.pipe(
   tap(x => { latestValue = x; })
).subscribe(x => {...});
```

## Reactively
You have several options here based on how you want to combine `Observables`.
For example when you have another `Observable` that emits mouse clicks and you want to have the latest value from your other `Observable` you could do something like
```ts
mouseclicks$.pipe(
   withLatestFrom(takeFourNumbers)
).subscribe(([mouseClick, latestNumber]) => {...})
```


## share/replay
You could also adapt the Observable to emit the latest value on subscription using `shareReplay`.
```ts
const takeFourNumbers = numbers.pipe(shareReplay(1), take(4));
```
There are some caveats though to `shareReplay`

## BehaviorSubject / ReplaySubject
A `ReplaySubject` has a buffer of last emitted values that will be emitted on subscription. This is very similar to `shareReplay`. You can't get this values *imperatively*.
A `BehaviorSubject` is similar to a `ReplaySubject` in that its buffer has the size of 1 and needs an initial value - so you can imagine it to always have something like a "current value". Upon subscription it will emit this current value or you can use it *imperatively* with `.getValue()`. 

Either way, in your example you would need connect your `Observable` to such a subject.

```ts
const subject = new ReplaySubject(1); // No initial value and we're only interested in the latest
const numbers = interval(1000);
const takeFourNumbers = numbers.pipe(take(4));
takeFourNumbers.subscribe(subject); // connects the Observable and the subject

subject.subscribe(x => console.log('Next: ', x));
setTimeout(()=>{
 console.log('how can we get the latest value which is 1?');
 subject.pipe(first()).subscribe(v=>console.log(v,'actual get'))
},2500)
```


I hope that this is of any help to you!
